### PR TITLE
Try timelapse support

### PR DESCRIPTION
**Pull request by Cryaaa**
This PR implements functionality allowing for the processing of time-lapse data. Firstly, to allow the use of large datasets I would suggest loading in dask stacks into napari (either manually or with the load stack option napari provides). 

Measurement can be performed on time lapse images now, producing a new column named 'frame'. This column is needed to identify which time point the values in the table belong to and if you want to upload your own csv files this column needs to be present for timelapse functionality!

The plotting is now dynamic meaning the points vary in size and alpha depending on the total amount of datapoints. Furthermore datapoints of the current frame are highlighted. This is one of many options we could implement for visualisation, so if there are any other suggestions I would love to hear some!

(Since processing of many cells/ objects can cause napari to freeze this branch already has multithreading implemented, so multithreading should be merged before this branch)

@haesleinhuepf , @zoccoler, @jo-mueller: No rush but if you have some time or want to try the cluster plotter with timelapse data please give me some feedback on if everything works or if there are issues / improvements to be made

### CODE REVIEW

**Comment by jo-mueller** [comment]
@Cryaaa could you somehow pass some data to play around with it?

**Comment by haesleinhuepf** [comment]
Hey @jo-mueller ,

https://github.com/clEsperanto/clesperanto_example_data/blob/main/CalibZAPWfixed_000154_max.tif

this is a timelapse dataset. You can use Voronoi-Otsu-Labeling to segment/label it and napari-skimage-regrgionprops to do measurements over time.

**Comment by haesleinhuepf** [comment]
https://zenodo.org/record/4276076#.YlZeBVNNZio

And this is a 3D+time dataset. Top-Hat-Backgroind+Subtraction and Voronoi-Otsu-Labeling work well in this case

**Comment by jo-mueller** [comment]
@Cryaaa would it be possible to update from the main branch? It seems to work but I am having a few NaNs in my measurements list and can't proceed through the dimensionality reduction :/

**Comment by Cryaaa** [comment]
@jo-mueller,
Sure! I thought the branch was already updated to the most recent main branch but I guess not! I'll let you know as soon as I've merged it :)

**Comment by Cryaaa** [comment]
@jo-mueller, weirdly if I update from main it doesn't implement the catch nans so I will have to do that manually...

**Comment by Cryaaa** [comment]
@jo-mueller 
the latest commit should add the nan functionality again and (hopefully) not break anything. If you find any weird behaviour just let me know but before I merged I looked through the files and am pretty sure that nothing should be affected by the changes being made....

**Comment by jo-mueller** [comment]
Yikes
```
=============================================== short test summary info ===============================================
FAILED napari_clusters_plotter/_tests/test_clustering.py::test_kmeans_clustering - ValueError: Shape of passed values...
FAILED napari_clusters_plotter/_tests/test_clustering.py::test_hdbscan_clustering - ValueError: Shape of passed value...
FAILED napari_clusters_plotter/_tests/test_clustering.py::test_gaussian_mixture_model - ValueError: Shape of passed v...
FAILED napari_clusters_plotter/_tests/test_clustering.py::test_agglomerative_clustering - ValueError: Shape of passed...
FAILED napari_clusters_plotter/_tests/test_clustering.py::test_mean_shift - ValueError: Shape of passed values is (2,...
FAILED napari_clusters_plotter/_tests/test_dimension_reduction.py::test_call_to_function
FAILED napari_clusters_plotter/_tests/test_dimension_reduction.py::test_umap - TypeError: umap() missing 1 required p...
FAILED napari_clusters_plotter/_tests/test_dimension_reduction.py::test_tsne - TypeError: tsne() missing 1 required p...
===================================== 8 failed, 10 passed, 25 warnings in 23.42s ======================================
```

**Comment by Cryaaa** [comment]
@jo-mueller 
yeah so there are a few bugs that arise when we implement the catch nans function unfortunately but I am working on it :D. I guess you'll have to wait a bit to try this branch....

**Comment by jo-mueller** [comment]
Cool! I can also take a look at it. I assume the problem is that we are having several columns in the dataframes that refer to similar things (`Label`, `Frame`, `Index`) and putting the data back together only operates on one of these criteria.

**Comment by Cryaaa** [comment]
@jo-mueller 
this could be one of the possible causes of the failing tests but other causes were already discussed in #44 and I think similar reasons are now present here. Maybe we have to merge the tests of that branch into here to get further but this might also be something we should discuss when Laura has some more free time to give some input as I have the feeling she has some more expertise here. Thanks for the suggestions and contributions though :)

**Comment by Cryaaa** [comment]
@jo-mueller and possibly @lazigu,
I fixed most of the tests but in the test_dimensionality reduction there is a call to qtbot although this function/object is not imported or defined. I found out that it is part of the pytestqt library but this library is not a part of the requirements in our setup. My question is how to deal with this error or where we have to specify the requirement for this function/object.

**Comment by lazigu** [comment]
Hi, Ryan @Cryaaa, I was just about to type you a message also! :D From today I am trying to get back into work bit by bit so if you need me I am here from now on. And I fixed tests locally of the progress_bar branch (just Github Cl is making a fuss still). If I understand right, there are no new tests on this branch, so it´s the same fixes as for the progress_bar one? 
For qtbot you don´t need to specify anything, pytest-qt is part of pytest so it just works like it is used now

**Comment by Cryaaa** [comment]
Hey @lazigu!
Glad to hear that you are recovering first of all :). If you have managed to fix the tests on that branch I will try and implement those changes in this branch too. Although thinking about it if the progress bar changes are functioning on that branch we could merge them into main before attempting to fix this branch. Anyway thans for the quick reply and don't let the github notifications stress you!

**Comment by Cryaaa** [comment]
@lazigu that sounds great! Then fingers crossed everything passes the tests :)

File: napari_clusters_plotter/_clustering.py [hunk 1]
```diff
@@ -575,7 +575,7 @@ def result_of_clustering(returned):
 @catch_NaNs
 def mean_shift(
     reg_props: pd.DataFrame, quantile: float = 0.2, n_samples: int = 50
-) -> (str, np.ndarray):
+) -> tuple[str, np.ndarray]:
```
Comment (lazigu) [inline napari_clusters_plotter/_clustering.py:578]: I had it also this type annotation once like this, but it gives TypeError in Github Cl, so you need to revert it back how it was 😅

Comment (jo-mueller) [reply napari_clusters_plotter/_clustering.py:578]: Hey both,

you can use Tuples as return type annotation, *but* you have to import it specifically!

To use tuples for returns, you'd have to:

```Python
from typing import Tuple

def function(...) -> Tuple(int, float):
```

Comment (lazigu) [reply napari_clusters_plotter/_clustering.py:578]: Ahh thanks, I learned something new!

Comment (Cryaaa) [reply napari_clusters_plotter/_clustering.py:578]: thanks so much @jo-mueller! That seems like the better solution.

Comment (jo-mueller) [reply napari_clusters_plotter/_clustering.py]: You're welcome. Just as a side note: You can also returns lists like this with 

```
from typing import List

def function() -> List[]:
```

I learned about that just the other day [here](https://stackoverflow.com/questions/40181344/how-to-annotate-types-of-multiple-return-values) - but it does not work with Napari  ^^" If you wanted to return multiple layers, you'd always have to use `->List[LayerDataTuple]`

**Comment by lazigu** [comment]
By the way, I think I would cherry-pick these your last couple of commits to the main branch, so these type annotation fixes are already in the new release :D

**Comment by lazigu** [comment]
@Cryaaa as in my case, on macos the job was buffering for 45min so I cancelled it, rerun failed one and it finished succesfully in 10min, so I guess for now we will have to keep doing this because I have no idea why it happens..

**Comment by haesleinhuepf** [comment]
I just wanted to let you all know that I tested this quite excessively and I love it!

umap_over_time

What keeps us from merging at the moment?

**Comment by jo-mueller** [comment]
> What keeps us from merging at the moment?

I just tried it with some "regular" non-timelapse data and receive an error in the plotting widget. Investigating!

**Comment by Cryaaa** [comment]
@haesleinhuepf,
One thing we might want to think about is the dependancy on pyclesperanto for Timelapse images but maybe we can just discuss this in the casual meeting? Johannes, Marcelo and I had a few ideas regarding this

**Comment by haesleinhuepf** [comment]
Yes, sure! I'm just amazed

umap_over_time2

**Comment by Cryaaa** [comment]
> > What keeps us from merging at the moment?
> 
> I just tried it with some "regular" non-timelapse data and receive an error in the plotting widget. Investigating!
> 
> Edit: The measurement/dimensionality reduction/clustering work fine.
> 
> Edit2: Here's the traceback. My first guess would be that the plotter widget is looking for a column called "frame" in the list of measurements, which does not necessarily exist for non-timelapse data.
> 
> ```
> File e:\biapol\projects\napari-clusters-plotter\napari_clusters_plotter\_plotter.py:629, in PlotterWidget.run(self=<napari_clusters_plotter._plotter.PlotterWidget object>, features=      label   centroid_x  centroid_y  centroid_z...1 -52.242638 -12.172477
> 
> [6881 rows x 16 columns], plot_x_axis_name='t-SNE_0', plot_y_axis_name='t-SNE_1', plot_cluster_name='', redraw_cluster_image=True)
>     625         self.viewer.layers.selection.add(s)
>     627 else:
>     628     a, sizes, colors_plot = unclustered_plot_parameters(
> --> 629         frame_id=features[POINTER].tolist(),
>         features =       label   centroid_x  centroid_y  centroid_z  ...  mean_max_distance_to_centroid_ratio  index    t-SNE_0    t-SNE_1
> 0         1     0.500000    9.833334         0.0  ...                             1.763667      1   7.672644  64.652878
> 1         2     0.000000   51.500000         0.0  ...                             1.000000      2 -61.477592 -36.015015
> 2         3     0.000000   65.000000         0.0  ...                                  NaN      3        NaN        NaN
> 3         4     0.600000  180.800003         0.0  ...                             1.751892      4  19.362715  54.895035
> 4         5     0.916667  234.166672         0.0  ...                             1.630703      5  19.264290  55.415401
> ...     ...          ...         ...         ...  ...                                  ...    ...        ...        ...
> 6876   6877  1295.000000  495.500000         0.0  ...                             1.000000   6877 -62.421459 -25.309450
> 6877   6878  1295.000000  531.500000         0.0  ...                             1.000000   6878 -62.173176 -25.365707
> 6878   6879  1295.000000  708.000000         0.0  ...                             1.500000   6879 -15.331562 -59.484077
> 6879   6880  1295.000000  906.000000         0.0  ...                                  NaN   6880        NaN        NaN
> 6880   6881  1295.000000  908.500000         0.0  ...                             1.000000   6881 -52.242638 -12.172477
> 
> [6881 rows x 16 columns]
>         POINTER = 'frame'
>         self = <napari_clusters_plotter._plotter.PlotterWidget object at 0x0000028F80014E50>
>         self.frame = 494
>         number_of_points = 6881
>     630         current_frame=self.frame,
>     631         n_datapoints=number_of_points,
>     632     )
>     634     # Potting
>     635     self.graphics_widget.pts = self.graphics_widget.axes.scatter(
>     636         self.data_x,
>     637         self.data_y,
>    (...)
>     640         alpha=a,
>     641     )
> 
> File ~\anaconda3\envs\devbio-napari-env\lib\site-packages\pandas\core\frame.py:3505, in DataFrame.__getitem__(self=      label   centroid_x  centroid_y  centroid_z...1 -52.242638 -12.172477
> 
> [6881 rows x 16 columns], key='frame')
>    3503 if self.columns.nlevels > 1:
>    3504     return self._getitem_multilevel(key)
> -> 3505 indexer = self.columns.get_loc(key)
>         key = 'frame'
>         indexer = None
>         self =       label   centroid_x  centroid_y  centroid_z  ...  mean_max_distance_to_centroid_ratio  index    t-SNE_0    t-SNE_1
> 0         1     0.500000    9.833334         0.0  ...                             1.763667      1   7.672644  64.652878
> 1         2     0.000000   51.500000         0.0  ...                             1.000000      2 -61.477592 -36.015015
> 2         3     0.000000   65.000000         0.0  ...                                  NaN      3        NaN        NaN
> 3         4     0.600000  180.800003         0.0  ...                             1.751892      4  19.362715  54.895035
> 4         5     0.916667  234.166672         0.0  ...                             1.630703      5  19.264290  55.415401
> ...     ...          ...         ...         ...  ...                                  ...    ...        ...        ...
> 6876   6877  1295.000000  495.500000         0.0  ...                             1.000000   6877 -62.421459 -25.309450
> 6877   6878  1295.000000  531.500000         0.0  ...                             1.000000   6878 -62.173176 -25.365707
> 6878   6879  1295.000000  708.000000         0.0  ...                             1.500000   6879 -15.331562 -59.484077
> 6879   6880  1295.000000  906.000000         0.0  ...                                  NaN   6880        NaN        NaN
> 6880   6881  1295.000000  908.500000         0.0  ...                             1.000000   6881 -52.242638 -12.172477
> 
> [6881 rows x 16 columns]
>    3506 if is_integer(indexer):
>    3507     indexer = [indexer]
> 
> File ~\anaconda3\envs\devbio-napari-env\lib\site-packages\pandas\core\indexes\base.py:3623, in Index.get_loc(self=Index(['label', 'centroid_x', 'centroid_y', 'cen...ex', 't-SNE_0', 't-SNE_1'],
>       dtype='object'), key='frame', method=None, tolerance=None)
>    3621     return self._engine.get_loc(casted_key)
>    3622 except KeyError as err:
> -> 3623     raise KeyError(key) from err
>         key = 'frame'
>    3624 except TypeError:
>    3625     # If we have a listlike key, _check_indexing_error will raise
>    3626     #  InvalidIndexError. Otherwise we fall through and re-raise
>    3627     #  the TypeError.
>    3628     self._check_indexing_error(key)
> 
> KeyError: 'frame'
> ```

Good that you spotted this. It seems like I programmed most of the new plotting features without thinking about non-timelapse data. I think that quite a lot of changes have to be made to make it work with both timelapse and non-timelapse data. So @haesleinhuepf: merging should come after this is fixed. If you want to take a look at it feel free but it might make more sense that I fix it since I also wrote most of the code for the changes to the plotter..

**Comment by haesleinhuepf** [comment]
> So @haesleinhuepf: merging should come after this is fixed. If you want to take a look at it feel free but it might make more sense that I fix it since I also wrote most of the code for the changes to the plotter..

Ok cool, please go ahead and fix the issues. Not super urgent, but high priority. Maybe there's time to finish this in the coming week?

**Comment by Cryaaa** [comment]
@haesleinhuepf 
I'm just looking at the code and I think it might be simpler than I thought. Do you think a timelapse check box would be alright? Because then it will be done by monday most likely :)

**Comment by haesleinhuepf** [comment]
The checkbox might not be necessary.

```
If len(image.shape) == 4: 
    #timelapse
```

This is how other tools do this differentiation... Would that work here?

**Comment by Cryaaa** [comment]
@haesleinhuepf this works for 3D timelapse data but would limit it to 3D. I think having the option of 2D timelapse data would also be nice. Although we could say that 2D timelapse has to have shape [t,z=1,y,x]. But yeah that would be the easy fix.

**Comment by haesleinhuepf** [comment]
Napari defines a dataset with 3 dimensions as 3D. It visualizes these three dimensions as spatial dimensions.

> Although we could say that 2D timelapse has to have shape [t,z=1,y,x]. But yeah that would be the easy fix.

I'd say a 2D-timelapse has to have shape `[t, 1, y, x]`. The [napari-time-slicer](https://www.napari-hub.org/plugins/napari-time-slicer) contains this section to explain this to users and developers:


_If you want to convert a 3D dataset into a 2D + time dataset, use the menu Tools > Utilities > Convert 3D stack to 2D timelapse (time-slicer). It will turn the 3D dataset to a 4D datset where the Z-dimension (index 1) has only 1 element, which will in napari be displayed with a time-slider. Note: It is recommended to remove the original 3D dataset after this conversion._


I'm just suggesting to implement timelapse-support in plugins such as napari-time-slicer and/or [napari-skimage-regionprops](https://github.com/haesleinhuepf/napari-skimage-regionprops/pull/20/files#diff-920a20c2bd609ff92772bc9c5c7718946325a8879b5a43da405bca2e713e2ffbR12) because then other plugins can reuse functionality. You better don't invent your own standard that only works with the clusters-plotter.

**Comment by Cryaaa** [comment]
@haesleinhuepf,
I remember you talking about something like this before so we'll go with that. I would say we just mention how timelapse data should look like for 2D datasets and then link to the time slicer or the napari-skimage-regionprops. I added a small function to utilities just for my testing which converts to this standard but I wasn't planning on adding it publicly to the clusters plotter :). Also one question while we are already chatting: Should I merge this and afterwards do the documentation for all functions which have changes or vice versa? I am asking because now that tests are also not randomly failing anymore I would love to merge and get ready for a release for next week.

**Comment by haesleinhuepf** [comment]
Two options: 
* Either you merge after adding documentation, 
* or you create an issue "Documentation missing about a,b,c" and merge.

Goal is just that we don't forget about the documentation.

**Pull request merged**